Voyageai refactoring and multimodal capability

fzowl opened the pull request:
Voyageai refactoring:
 - contextual model(s)
 - remove default model value
 - token counting and effective batching
 - test with new reranker, rerank-2.5 as well
 - multimodal capability (text, image, video)
 - more tests

<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> **Medium Risk**
> Updates VoyageAI embedding behavior and public surface (new deprecations for implicit `model` and `batch_size`, plus new token-counting APIs), which could affect existing integrations and batching behavior. Most risk is functional/regression around embedding calls and new integration tests that depend on external tooling (API keys, Pillow, ffmpeg).
> 
> **Overview**
> Improves `VoyageAIVectorizer` by adding local token counting (`count_tokens`/`acount_tokens`) and introducing a `VOYAGE_TOTAL_TOKEN_LIMITS` table to support token-aware batching.
> 
> Deprecates instantiating the vectorizer without an explicit `model` (now emits a `DeprecationWarning` and will remove the implicit default in a future major version) and deprecates passing `batch_size` to `_embed_many`/`_aembed_many` in favor of automatic batch sizing.
> 
> Expands integration coverage with VoyageAI-specific tests for token counting, token limit constants, and multimodal embedding flows (text, image, video, and async), including conditional skipping when `ffmpeg` isn’t available.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 1e971a6a93ee5078641175e81808f6a1dea5aeab. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

fzowl commented on redisvl/utils/vectorize/text/voyageai.py:
```diff
@@ -329,25 +372,135 @@ async def _aembed_many(
+        try:
+            token_counts = self.count_tokens(texts)
+        except Exception as e:
+            raise ValueError(f"Failed to count tokens for batching: {e}")
```
count_token() is a low-latency local function.

fzowl commented:
@bsbodden Yes, the contextualized_embed functions were introduced in the voyageai package 0.3.5 version, that's why i changed it in the `pyproject.toml`.

fzowl committed:
Add token counting and context model detection to VoyageAI

fzowl committed:
Add multimodal tests for VoyageAI (text, image, video, async)

vishal-bala reviewed (commented):
Hey @fzowl, thanks for updating this PR to build on the latest vectorizer framework (and sorry for the delay reviewing) - this looks good to me. Since you made the changes, VoyageAI introduced their 4th generation models (`voyage-4`, etc.) - could you add those to your token limit mapping?

Also, could you rebase your PR off the current `main` branch for RedisVL? We've introduced some improvements to our CI/CD to allow our testing workflows to be runnable on PRs from forked repositories!

fzowl committed:
Address review feedback: deprecation warnings and documentation

- Keep default model with DeprecationWarning when not explicitly set
- Add runtime DeprecationWarning when deprecated batch_size is passed
- Clarify count_tokens() is local/CPU-only in docstrings

vishal-bala commented on redisvl/utils/vectorize/voyageai.py:
```diff
@@ -13,6 +14,29 @@
+_MODEL_NOT_SET = object()
+
+# Token limits for VoyageAI models (used for token-aware batching)
+VOYAGE_TOTAL_TOKEN_LIMITS = {
```
Can we add the `voyage-4` family of models here too?

vishal-bala commented on redisvl/utils/vectorize/voyageai.py:
```diff
@@ -336,8 +389,17 @@ def _embed_many(
+                DeprecationWarning,
+                stacklevel=2,
+            )
+        else:
```
^ This is a valid comment. If anything, we should implement basic logic to warn once and only once.

vishal-bala commented:
Thanks for the continued work here. I found a couple of follow-up items that look worth tightening before merge:

1. Multimodal `embed_many` appears to collapse multiple requested contents into one multimodal input. For multimodal models, `embed_many(["Ocean waves", "Forest trees"])` looks like it will send `[batch]`, which VoyageAI interprets as a single multimodal input with two text parts, rather than two independent inputs. That means the API can return one embedding for two requested contents, breaking the `embed_many` contract and also causing cache miss insertion to misalign. I think the multimodal branch should preserve one outer item per requested embedding, e.g. wrap each item as its own multimodal input, and the sync/async tests should assert the returned length matches the requested content count.

2. Nice addition exposing local token counting. For multimodal models, though, text token counts alone will not capture the actual request usage because images/videos contribute pixel-derived tokens in VoyageAI accounting. Could we either make it explicit that `count_tokens()` is text-only, or add a separate multimodal usage path using VoyageAI `count_usage()` for image/video inputs? This does not need to change the text-only path, but it would help avoid users assuming token-aware batching covers multimodal payloads when it currently only has enough information for text.

vishal-bala reviewed (commented):
A couple of smaller items too 😄
